decrypt.go

```diff
@@ -80,11 +82,27 @@ func (ds *decryptStream) read(b []byte) (n int, err error) {
 
 		if last {
 			ds.state = stateEndOfStream
+			// If we've reached the end of the stream, but
+			// have data left (which only happens in V2),
+			// return so that the next call(s) will hit
+			// the case at the top, and then we'll hit the
+			// case below.
+			if len(ds.buf) > 0 {
+				if ds.version.Major < 2 {
+					panic(fmt.Sprintf("version=%s, last=true, len(ds.buf)=%d > 0", ds.version, len(ds.buf)))
+				}
+
+				return n, nil
+			}
 		}
 	}
 
 	if ds.state == stateEndOfStream {
 		ds.err = assertEndOfStream(ds.mps)
+		// If V2, we can hit EOF with n > 0.
+		if ds.err == io.EOF {
+			return n, ds.err
+		}
 		if ds.err != nil {
 			return 0, ds.err
```

Comment:
If we wrote out a non-zero number of plaintext bytes, we need to report that accurately even if we hit some crazy error, right? Maybe there shouldn't be any branch here at all, just `return n, ds.err`?